ride/cf-staking.ride

```diff
@@ -0,0 +1,283 @@
+}
+
+#both
+@Callable(i)
+func withdrawFarmTokens (amount: Int, compound: Boolean) = {
+  let shareTokenId = getShareAssetId()
+  if compound
+  then {
+    #Calculate vp price for 1 farm token to be able to payout
+    let staked = tryGetInteger(keyStakedCompound)
+    let total = tryGetInteger(keyTotalCompound)
+    let vp = calcVirtualPrice(staked,total)
+
+    #Get amount of staked
+    let keyStakedCompoundU = keyStakedCompoundUser(i.caller)
+    let personalStaked = tryGetInteger(keyStakedCompoundU)
+
+    #Calculate the virtual amount from your tokens to wd
+    let virtualWd = if (amount == -1)then personalStaked else fraction(amount, VPScale, vp)
+    let amountWd =  if (amount == -1) then fraction(virtualWd,vp,VPScale)else amount
+
+    if virtualWd > personalStaked
+    then throw("You don't have so much funds to withdraw")
+    else{
+      [
+          IntegerEntry(keyStakedCompoundU, personalStaked-virtualWd),
+          IntegerEntry(keyStakedCompound, staked-virtualWd),
+          IntegerEntry(keyTotalCompound, total-amountWd),
+          ScriptTransfer(i.caller,amountWd,shareTokenId),
+          IntegerEntry(keyLastVP,vp),
+          IntegerEntry("last_virt_compound_wd",virtualWd),
+          IntegerEntry("last_amount_compound_wd",amountWd)
+      ]
+    }
+  }
+  else{
+    let addressStr = toString(i.caller)
+    let stakedAmount = tryGetInteger(keyStakedAmount(i.caller))
+    let wdAmount = if (amount == -1) then stakedAmount else amount
+    if ((wdAmount > stakedAmount))
+        then throw("you don't have tokens available")
+        else (
+          claimStakingResult(i.caller) ++ 
+          [
+            IntegerEntry(keyStakedAmount(i.caller), (stakedAmount - wdAmount)), 
+            IntegerEntry(keyGlobalStaked, (tryGetInteger(keyGlobalStaked) - wdAmount)), 
+            ScriptTransfer(i.caller, wdAmount, shareTokenId),
+            IntegerEntry("last_staking_wd",wdAmount)
+          ]
+        )
+  }
+
+}
+
+#both
+@Callable(i)
+func stakeFarmTokens (compound: Boolean) = {
+  let shareTokenId = getShareAssetId()
+  if (i.payments[0].assetId != shareTokenId)
+  then throw("Wrong assetId")
+  else{
+    if compound
+    then {
+      let isLocked = isCollectiveFarmLocked()
+      if isLocked 
+      then throw("The farm is closed for investments, it can't compound")
+      else {
+        #get amount of already staked funds
+        let keyStakedCompoundU = keyStakedCompoundUser(i.caller)
+        let alreadyStaked = tryGetInteger(keyStakedCompoundU)
+
+        #get total amount funds
+        let total = tryGetInteger(keyTotalCompound)
+        #get total amount userstaked funds
+        let staked = tryGetInteger(keyStakedCompound)
+        #calculate exactStakeAmount after normalization
+        let vp = calcVirtualPrice(staked,total)
+        let rawStakeAmount = i.payments[0].amount
+        let exactStakeAmount = fraction(rawStakeAmount, VPScale, vp)
+
+        #bug fix to make sure first claim also succeeds
```

Comment:
Why is it here? I do not like this workaround.

Reply:
The comment might be misplaced actually, or do you mean the code/logic could be improved?

Reply:
What `fix to make sure first claim also succeeds` problem it is? I do not get it at all.